From 3ad713469691dda650f324f9eead9714683eaff3 Mon Sep 17 00:00:00 2001
From: Arun Raghavan <arunsr@codeaurora.org>
Date: Tue, 29 May 2018 09:03:52 +0530
Subject: [PATCH 2/3] audiobasesrc: Handle capture of non-PCM data

Just need to make sure that we're not assuming the existence of a
channel count or sample format while fixating caps.
---
 gst-libs/gst/audio/gstaudiobasesrc.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/gst-libs/gst/audio/gstaudiobasesrc.c b/gst-libs/gst/audio/gstaudiobasesrc.c
index e8063eb97..ce2522a01 100644
--- a/gst-libs/gst/audio/gstaudiobasesrc.c
+++ b/gst-libs/gst/audio/gstaudiobasesrc.c
@@ -497,11 +497,15 @@ gst_audio_base_src_fixate (GstBaseSrc * bsrc, GstCaps * caps)
 
   s = gst_caps_get_structure (caps, 0);
 
-  /* fields for all formats */
+  /* rate is expected on all caps, channels and format are expected for PCM
+   * (audio/x-raw) */
   gst_structure_fixate_field_nearest_int (s, "rate", GST_AUDIO_DEF_RATE);
-  gst_structure_fixate_field_nearest_int (s, "channels",
-      GST_AUDIO_DEF_CHANNELS);
-  gst_structure_fixate_field_string (s, "format", GST_AUDIO_DEF_FORMAT);
+
+  if (gst_structure_has_field (s, "channels"))
+    gst_structure_fixate_field_nearest_int (s, "channels",
+        GST_AUDIO_DEF_CHANNELS);
+  if (gst_structure_has_field (s, "format"))
+    gst_structure_fixate_field_string (s, "format", GST_AUDIO_DEF_FORMAT);
 
   caps = GST_BASE_SRC_CLASS (parent_class)->fixate (bsrc, caps);
 
-- 
2.18.0

